### ISSUE TITLE
non-reproducibility of GPU redrock

### ISSUE THREAD

**Issue by sbailey** [desihub/redrock#242]
When run in GPU mode, redrock does not produce identical output even when run back-to-back on the same node using the same inputs.  The differences are small enough that they don't matter for final science analyses, but they are a hassle when evaluating PRs etc because you can't simply use fitsdiff to confirm that the outputs didn't change.

Example redrock runs on GPU and CPU:
```
cd $CFS/desi/spectro/redux/iron/tiles/cumulative/100/20210505

srun -n 64 -c 2 --gpu-bind=map_gpu:3,2,1,0 --cpu-bind=cores \
    rrdesi_mpi --gpu -i coadd-0-100-thru20210505.fits -o $SCRATCH/redrock-gpu-A.fits
    
srun -n 64 -c 2 --gpu-bind=map_gpu:3,2,1,0 --cpu-bind=cores \
    rrdesi_mpi --gpu -i coadd-0-100-thru20210505.fits -o $SCRATCH/redrock-gpu-B.fits

srun -n 64 -c 2 --gpu-bind=map_gpu:3,2,1,0 --cpu-bind=cores \
    rrdesi_mpi -i coadd-0-100-thru20210505.fits -o $SCRATCH/redrock-cpu-C.fits
    
srun -n 64 -c 2 --gpu-bind=map_gpu:3,2,1,0 --cpu-bind=cores \
    rrdesi_mpi -i coadd-0-100-thru20210505.fits -o $SCRATCH/redrock-cpu-D.fits
```

Differences evaluated with
```python
from astropy.table import Table
import numpy as np
from os.path import expandvars

#- GPU
a = Table.read(expandvars('$SCRATCH/redrock-gpu-A.fits'), 'REDSHIFTS')
b = Table.read(expandvars('$SCRATCH/redrock-gpu-B.fits'), 'REDSHIFTS')

#- CPU
c = Table.read(expandvars('$SCRATCH/redrock-cpu-C.fits'), 'REDSHIFTS')
d = Table.read(expandvars('$SCRATCH/redrock-cpu-D.fits'), 'REDSHIFTS')

print('GPU re-run consistency')
for col in ('Z', 'ZERR', 'CHI2'):
    ndiff = np.sum(a[col] != b[col])
    std = np.std(a[col]-b[col])
    print(f'{col=}  {ndiff=}  {std=}')

print('CPU re-run consistency')
for col in ('Z', 'ZERR', 'CHI2'):
    ndiff = np.sum(c[col] != d[col])
    std = np.std(c[col]-d[col])
    print(f'{col=}  {ndiff=}  {std=}')
```

Results:
```
GPU re-run consistency
col='Z'  ndiff=85  std=8.807669764011719e-17
col='ZERR'  ndiff=107  std=8.347450031031658e-18
col='CHI2'  ndiff=92  std=1.5223154341921055e-07

[corrected from original post were SB pasted the wrong snippet]
CPU re-run consistency
col='Z'  ndiff=0  std=0.0
col='ZERR'  ndiff=0  std=0.0
col='CHI2'  ndiff=0  std=0.0
```

Although all 3 of Z, ZERR, and CHI2 are float64 and come from the same parabola fit to a chi2 vs. z scan, Z and ZERR vary by double precision roundoff errors while CHI2 varies by single precision roundoff errors.

It would helpful to know exactly where the differences arise, why, and whether there is a magic environment variable or parameter we could use to enforce reproducibility.

In the past, MKL has claimed that it is a feature not a bug that MKL linear algebra is not bitwise reproducible.  Paraphrasing: it's faster and you shouldn't care about round off precision bits.  I suspect there is something like that going on with cupy.

@craigwarner-ufastro @dmargala and heads up @lastephey @akremin @marcelo-alvarez

**Comment by craigwarner-ufastro** [comment]
I've been able to reproduce and confirm Stephen's results.  When running twice in a row with identical inputs, there are rounding errors in the GPU results with respect to each other (and to the CPU results - which was already known).  The CPU results are identical so correcting a copy/paste error above:

```
GPU re-run consistency
col='Z'  ndiff=70  std=6.962342309921365e-17
col='ZERR'  ndiff=100  std=1.6775324122317275e-17
col='CHI2'  ndiff=82  std=3.0689534932914646e-07
```
```
CPU re-run consistency
col='Z'  ndiff=0  std=0.0
col='ZERR'  ndiff=0  std=0.0
col='CHI2'  ndiff=0  std=0.0
```
```
CPU vs GPU
col='Z'  ndiff=419  std=6.080079552416836e-15
col='ZERR'  ndiff=452  std=1.5040341339977628e-13
col='CHI2'  ndiff=420  std=6.021641708067369e-07
```
I suspect that Stephen is right that it is something going on with CuPy but I will investigate further to attempt to track down where it is coming from and why the magnitude is different for CHI2.

**Comment by craigwarner-ufastro** [comment]
I have tracked down the origin of the GPU discrepancies in `fitz.py` to the function call

```
        zmin, sigma, chi2min, zwarn = minfit(zz[i-1:i+2], zzchi2[i-1:i+2])
```
And specifically within minfit, the call to `a,b,c = np.polyfit(x,y,2)`.  I have no idea why np.polyfit is introducing errors, but I was able to confirm by running with a small ntargets and adding print statements that this is where the run-to-run differences are appearing.  

It is also a phenomenon that I can characterize as the same targets usually show an oscillating discrepancy.  For instance, if I run the same example above with `--ntarget 10` for brevity, then it is always 1 and 6 and sometimes 3 or 5 that differ in CHI2:
```
>>> np.where(a['CHI2'] != b['CHI2'])
(array([1, 3, 6]),)
```
Moreover, running this example 4 times in a row, the 8 values I got for a['CHI2'][1] oscillated between 3 actual values: 7632.602427930571, 7632.602427926846, and 7632.6024279259145 -- four of the 8 runs produced the first value and two each produced the other two values (but in all 4 A/B pairs, one of them was ...571).  Very strange.

Going back to the source of the error, if I insert print statements in and around the call to `minfit`, I can see in two consecutive runs for the GALAXY template, for target ID 39633396149781440, both print out that the input `i`, `zz[i-1:i+2]`, and `zzchi2[i-1:i+2]` are the same.  But `a`, `b`, and `c` resulting from polyfit are slightly different, which results in different `zmin`, `sigma` (zerr), and `chi2min`.  The reason the error is bigger in `chi2min` is that it is calculated using b**2 in the minfit algorithm:
```
ID 39633396149781440 IMIN 151
ZZ [0.10439183 0.10450081 0.1046098 ] ZZCHI [9599.21471016 9555.87009839 9584.68495262]
A 3037660806.695476 B -634942700.2849765 C 33189052.681673795
i 8 ZMIN 0.10451178401575716 SIGMA 1.814388819837914e-05 CHI2MIN 9555.50439124927 ZWARN 0
```
versus
```
ID 39633396149781440 IMIN 151
ZZ [0.10439183 0.10450081 0.1046098 ] ZZCHI [9599.21471016 9555.87009839 9584.68495262]
A 3037660806.695524 B -634942700.2849865 C 33189052.681674328
i 8 ZMIN 0.10451178401575714 SIGMA 1.8143888198378997e-05 CHI2MIN 9555.504391260445 ZWARN 0
```

**Comment by sbailey** [comment]
FYI, from previous interactions with the Intel MKL team:

> Thank you for the detailed reproducer!  I was able to reproduce the behavior, but this is not a bug.  Intel MKL does not guarantee bit-wise identical results by default, as there may be a performance impact to do so.  However, Intel MKL does offer a conditional numerical reproducibility feature that will provide reproducible results, subject to some limitations.  For instance, there are some codepaths in Intel MKL that rely on aligned data, but if CNR mode is enabled then those codepaths are not taken regardless of data alignment.

The proposed workaround was to set `export MKL_CBWR=AUTO` which fixed the non-reproducibility on KNL CPUs.  @craigwarner-ufastro please try that for these test cases, including whether that has any noticable performance hit for us.

Other info on Intel/MKL's view of (non)reproducibility:

  * https://software.intel.com/content/www/us/en/develop/articles/introduction-to-the-conditional-numerical-reproducibility-cnr.html
  * https://software.intel.com/content/www/us/en/develop/documentation/mkl-linux-developer-guide/top/obtaining-numerically-reproducible-results/getting-started-with-conditional-numerical-reproducibility.html

**Comment by craigwarner-ufastro** [comment]
I set the env var `export MKL_CBWR=AUTO` and it results in a crash in the coarse redshift scan in both CPU and GPU mode with an `Illegal instruction` :(

```
cdwarner@nid001033:/global/cfs/cdirs/desi/users/cdwarner/code> cd $CFS/desi/spectro/redux/iron/tiles/cumulative/100/20210505
cdwarner@nid001033:/global/cfs/cdirs/desi/spectro/redux/iron/tiles/cumulative/100/20210505> export MKL_CBWR=AUTO
cdwarner@nid001033:/global/cfs/cdirs/desi/spectro/redux/iron/tiles/cumulative/100/20210505> srun -n 1 -c 2 --gpu-bind=map_gpu:0 --cpu-bind=cores     rrdesi_mpi -i coadd-0-100-thru20210505.fits -o $SCRATCH/redrock-cpu-C.fits --ntarget 10
Running with 1 processes
Loading targets...
Read and distribution of 10 targets: 0.2 seconds
DEBUG: Read templates from /global/common/software/desi/perlmutter/desiconda/20230111-2.1.0/code/redrock-templates/0.8
DEBUG: Using default redshift range -0.0050-1.6997 for rrtemplate-galaxy.fits
DEBUG: Using redshift range 1.4000-6.9930 for rrtemplate-qso-HIZ.fits
DEBUG: Using redshift range 0.0500-1.5983 for rrtemplate-qso-LOZ.fits
DEBUG: Using default redshift range -0.0020-0.0020 for rrtemplate-star-A.fits
DEBUG: Using default redshift range -0.0020-0.0020 for rrtemplate-star-B.fits
DEBUG: Using default redshift range -0.0020-0.0020 for rrtemplate-star-CV.fits
DEBUG: Using default redshift range -0.0020-0.0020 for rrtemplate-star-F.fits
DEBUG: Using default redshift range -0.0020-0.0020 for rrtemplate-star-G.fits
DEBUG: Using default redshift range -0.0020-0.0020 for rrtemplate-star-K.fits
DEBUG: Using default redshift range -0.0020-0.0020 for rrtemplate-star-M.fits
DEBUG: Using default redshift range -0.0020-0.0020 for rrtemplate-star-WD.fits
Read and broadcast of 11 templates: 0.0 seconds
Creating GPU context: 0.0 seconds
Rebinning templates: 5.5 seconds
Computing redshifts
  Scanning redshifts for template GALAXY
    Progress:   0 %
srun: error: nid001033: task 0: Illegal instruction
srun: launch/slurm: _step_signal: Terminating StepId=6618032.5
```

Note that the crash is not in `np.polyfit` but somewhere else before that in the initial coarse redshift scan.

**Comment by craigwarner-ufastro** [comment]
I have isolated the MKL crash - it is in the numpy @ operator, in the line `OIIflux = np.sum(zcoeff[j] @ OIItemplate.T, axis=1)`

I created a test.py that reproduces this error:
```
import numpy as np

x = np.ones((10,90))
y = np.ones((1446,10))
z = y@x
print ("OK")
```

```
cdwarner@nid001033:/global/cfs/cdirs/desi/users/cdwarner/code> export MKL_CBWR=AUTO
cdwarner@nid001033:/global/cfs/cdirs/desi/users/cdwarner/code> python test.py 
Illegal instruction
cdwarner@nid001033:/global/cfs/cdirs/desi/users/cdwarner/code> unset MKL_CBWR 
cdwarner@nid001033:/global/cfs/cdirs/desi/users/cdwarner/code> python test.py 
OK
```

**Comment by craigwarner-ufastro** [comment]
I tried Daniel's suggestion of using `MKL_CBWR=COMPATIBLE` and that does indeed avoid the crash for Illegal instruction; however we still have the same floating point rounding error differences in the output on successive GPU runs coming from np.polyfit.

I then tried on my machine using `MKL_CBWR=AUTO` because I noticed my conda environment has mkl=2021.4.0, a newer version and I confirm that this also does not crash (and I also have an AMD processor); however the floating point rounding errors on consecutive runs are still there as well.

**Comment by sbailey** [comment]
We've been pinning to MKL=2020.0 because that was the last version that was reliably reproducible on CPUs without requiring setting magic environment variables.  Specter linear algebra was where it bit us before.

As a hack for this particular case, what about rounding the outputs of `np.polyfit` before using them to calculate Z, ZERR, and CHI2?  i.e. drop the bits that MKL is being random about since that level of precision doesn't actually matter for us but the non-reproducibility is a hassle.

Detail: `numpy.round` is for a fixed number of decimal places; in this case I think we want to preserve a relative precision, i.e. number of total digits not number of decimal places so we'd need to be slightly clever.

**Comment by sbailey** [comment]
fixed in PR #245.  Closing.
